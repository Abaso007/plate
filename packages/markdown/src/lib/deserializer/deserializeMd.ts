import type { Descendant, SlateEditor } from '@udecode/plate';
import type { Root } from 'mdast';
import type { Plugin } from 'unified';

import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import remarkMdx from 'remark-mdx';
import remarkParse from 'remark-parse';
import { unified } from 'unified';

import type { TNodes } from '../nodesRule';

import { MarkdownPlugin } from '../MarkdownPlugin';
import { mdastToSlate } from './mdastToSlate';


// TODO: fixes tests
const remarkSplitLineBreaks: Plugin<[deserializeOptions?], Root, Root> = function ({ splitLineBreaks = false } = {}) {
  if (!splitLineBreaks) return;

  return function transformer(tree: Root) {
    const processNode = (node: any) => {
      if (!node.position) return node;

      if (node.children) {
        const newChildren: any[] = [];
        let lastLine = node.position.start.line;

        node.children.forEach((child: any, index: number) => {
          const isFirstChild = index === 0;
          const isLastChild = index === node.children.length - 1;

          // Add empty paragraphs for line breaks before the child
          const emptyLinesBefore = child.position.start.line - (isFirstChild ? lastLine : lastLine + 1);
          if (emptyLinesBefore > 0) {
            for (let i = 0; i < emptyLinesBefore; i++) {
              newChildren.push({
                children: [{ type: 'text', value: '' }],
                position: {
                  end: { column: 1, line: lastLine + i + 1 },
                  start: { column: 1, line: lastLine + i + 1 }
                },
                type: 'paragraph'
              });
            }
          }

          // Add the child node
          newChildren.push(processNode(child));

          // Add empty paragraphs for line breaks after the last child
          if (isLastChild) {
            const emptyLinesAfter = node.position.end.line - child.position.end.line - 1;
            if (emptyLinesAfter > 0) {
              for (let i = 0; i < emptyLinesAfter; i++) {
                newChildren.push({
                  children: [{ type: 'text', value: '' }],
                  position: {
                    end: { column: 1, line: child.position.end.line + i + 1 },
                    start: { column: 1, line: child.position.end.line + i + 1 }
                  },
                  type: 'paragraph'
                });
              }
            }
          }

          lastLine = child.position.end.line;
        });

        node.children = newChildren;
      }

      return node;
    };

    return processNode(tree);
  };
};

export const deserializeMd = (editor: SlateEditor, data: string) => {
  // if using remarkMdx, we need to replace <br> with <br /> since <br /> is not supported in mdx.
  data = data.replaceAll('<br>', '<br />');

  const { nodes, splitLineBreaks } = editor.getOptions(MarkdownPlugin);

  const toSlateProcessor = unified()
    .use(remarkParse)
    .use(remarkMdx)
    .use(remarkGfm)
    .use(remarkMath)
    .use(remarkSplitLineBreaks, { splitLineBreaks })
    .use(remarkToSlate, {
      nodes,
      splitLineBreaks,
    });

  return toSlateProcessor.processSync(data).result;
};

export type deserializeOptions = {
  nodes?: TNodes;
  splitLineBreaks?: boolean;
};

declare module 'unified' {
  interface CompileResultMap {
    remarkToSlateNode: Descendant[];
  }
}

const remarkToSlate: Plugin<[deserializeOptions?], Root, Descendant[]> =
  function ({ nodes, splitLineBreaks = false } = {}) {
    this.compiler = function (node) {
      return mdastToSlate(node as Root, { nodes, splitLineBreaks });
    };
  };
