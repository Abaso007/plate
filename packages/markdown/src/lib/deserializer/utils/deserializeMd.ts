import type { SlateEditor } from '@udecode/plate';

import remarkGfm from 'remark-gfm';
import remarkParse from 'remark-parse';
import { type Processor, unified } from 'unified';

import { MarkdownPlugin } from '../../MarkdownPlugin';
import {
  type RemarkElementRules,
  type RemarkPluginOptions,
  type RemarkTextRules,
  remarkPlugin,
} from '../../remark-slate';
import { parseMarkdownBlocks } from './parseMarkdownBlocks';

export type DeserializeMdOptions = {
  /** Token types to filter out from marked lexer */
  excludeTokens?: string[];
  /** Whether to add _memo property to elements */
  memoize?: boolean;
  /** A function that allows you to modify the markdown processor. */
  processor?: (processor: Processor) => Processor;
};

/** Deserialize content from Markdown format to Slate format. */
export const deserializeMd = (
  editor: SlateEditor,
  data: string,
  {
    excludeTokens = ['space'],
    memoize = true,
    processor,
  }: DeserializeMdOptions = {}
) => {
  const elementRules: RemarkElementRules = {};
  const textRules: RemarkTextRules = {};

  const options = editor.getOptions(MarkdownPlugin);

  Object.assign(elementRules, options.elementRules);
  Object.assign(textRules, options.textRules);

  let tree: any = unified().use(remarkParse);

  if (processor) {
    tree = processor(tree);
  }

  tree = tree.use(remarkGfm).use(remarkPlugin, {
    editor,
    elementRules,
    indentList: options.indentList,
    textRules,
  } as unknown as RemarkPluginOptions);

  if (memoize) {
    return parseMarkdownBlocks(data, { excludeTokens }).map((token) => ({
      _memo: token.raw,
      ...tree.processSync(token.raw).result[0],
    }));
  }

  return tree.processSync(data).result;
};

// TODO: Collect rules from plugins
// editor.plugins.forEach((plugin: SlatePlugin) => {
//   if (plugin.parsers?.markdown?.deserialize) {
//     const { elementRules: pluginElementRules, textRules: pluginTextRules } =
//       plugin.parsers.markdown.deserialize as MarkdownDeserializer;
//
//     if (pluginElementRules) {
//       Object.assign(elementRules, pluginElementRules);
//     }
//     if (pluginTextRules) {
//       Object.assign(textRules, pluginTextRules);
//     }
//   }
// });
