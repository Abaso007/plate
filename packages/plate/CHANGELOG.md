# @udecode/plate

## 42.0.0

### Major Changes

- [#3920](https://github.com/udecode/plate/pull/3920) by [@zbeyens](https://github.com/zbeyens) – **This package is now the new common package**, so all plugin packages are being removed. **Migration**:

  - Add the following dependencies:

  ```json
  "@udecode/plate-alignment": "42.0.0",
  "@udecode/plate-autoformat": "42.0.0",
  "@udecode/plate-basic-elements": "42.0.0",
  "@udecode/plate-basic-marks": "42.0.0",
  "@udecode/plate-block-quote": "42.0.0",
  "@udecode/plate-break": "42.0.0",
  "@udecode/plate-code-block": "42.0.0",
  "@udecode/plate-combobox": "42.0.0",
  "@udecode/plate-comments": "42.0.0",
  "@udecode/plate-csv": "42.0.0",
  "@udecode/plate-diff": "42.0.0",
  "@udecode/plate-docx": "42.0.0",
  "@udecode/plate-find-replace": "42.0.0",
  "@udecode/plate-floating": "42.0.0",
  "@udecode/plate-font": "42.0.0",
  "@udecode/plate-heading": "42.0.0",
  "@udecode/plate-highlight": "42.0.0",
  "@udecode/plate-horizontal-rule": "42.0.0",
  "@udecode/plate-indent": "42.0.0",
  "@udecode/plate-indent-list": "42.0.0",
  "@udecode/plate-kbd": "42.0.0",
  "@udecode/plate-layout": "42.0.0",
  "@udecode/plate-line-height": "42.0.0",
  "@udecode/plate-link": "42.0.0",
  "@udecode/plate-list": "42.0.0",
  "@udecode/plate-markdown": "42.0.0",
  "@udecode/plate-media": "42.0.0",
  "@udecode/plate-mention": "42.0.0",
  "@udecode/plate-node-id": "42.0.0",
  "@udecode/plate-normalizers": "42.0.0",
  "@udecode/plate-reset-node": "42.0.0",
  "@udecode/plate-resizable": "42.0.0",
  "@udecode/plate-select": "42.0.0",
  "@udecode/plate-selection": "42.0.0",
  "@udecode/plate-slash-command": "42.0.0",
  "@udecode/plate-suggestion": "42.0.0",
  "@udecode/plate-tabbable": "42.0.0",
  "@udecode/plate-table": "42.0.0",
  "@udecode/plate-toggle": "42.0.0",
  "@udecode/plate-trailing-block": "42.0.0"
  ```

  - Either replace all `@udecode/plate` imports with the individual package imports, or export the following in a new file (e.g. `src/plate.ts`):

  ```ts
  export * from '@udecode/plate-alignment';
  export * from '@udecode/plate-autoformat';
  export * from '@udecode/plate-basic-elements';
  export * from '@udecode/plate-basic-marks';
  export * from '@udecode/plate-block-quote';
  export * from '@udecode/plate-break';
  export * from '@udecode/plate-code-block';
  export * from '@udecode/plate-combobox';
  export * from '@udecode/plate-comments';
  export * from '@udecode/plate-diff';
  export * from '@udecode/plate-find-replace';
  export * from '@udecode/plate-font';
  export * from '@udecode/plate-heading';
  export * from '@udecode/plate-highlight';
  export * from '@udecode/plate-horizontal-rule';
  export * from '@udecode/plate-indent';
  export * from '@udecode/plate-indent-list';
  export * from '@udecode/plate-kbd';
  export * from '@udecode/plate-layout';
  export * from '@udecode/plate-line-height';
  export * from '@udecode/plate-link';
  export * from '@udecode/plate-list';
  export * from '@udecode/plate-media';
  export * from '@udecode/plate-mention';
  export * from '@udecode/plate-node-id';
  export * from '@udecode/plate-normalizers';
  export * from '@udecode/plate-reset-node';
  export * from '@udecode/plate-select';
  export * from '@udecode/plate-csv';
  export * from '@udecode/plate-docx';
  export * from '@udecode/plate-markdown';
  export * from '@udecode/plate-slash-command';
  export * from '@udecode/plate-suggestion';
  export * from '@udecode/plate-tabbable';
  export * from '@udecode/plate-table';
  export * from '@udecode/plate-toggle';
  export * from '@udecode/plate-trailing-block';
  export * from '@udecode/plate-alignment/react';
  export * from '@udecode/plate-autoformat/react';
  export * from '@udecode/plate-basic-elements/react';
  export * from '@udecode/plate-basic-marks/react';
  export * from '@udecode/plate-block-quote/react';
  export * from '@udecode/plate-break/react';
  export * from '@udecode/plate-code-block/react';
  export * from '@udecode/plate-combobox/react';
  export * from '@udecode/plate-comments/react';
  export * from '@udecode/plate-floating';
  export * from '@udecode/plate-font/react';
  export * from '@udecode/plate-heading/react';
  export * from '@udecode/plate-highlight/react';
  export * from '@udecode/plate-layout/react';
  export * from '@udecode/plate-slash-command/react';
  export * from '@udecode/plate-indent/react';
  export * from '@udecode/plate-indent-list/react';
  export * from '@udecode/plate-kbd/react';
  export * from '@udecode/plate-line-height/react';
  export * from '@udecode/plate-link/react';
  export * from '@udecode/plate-list/react';
  export * from '@udecode/plate-media/react';
  export * from '@udecode/plate-reset-node/react';
  export * from '@udecode/plate-selection';
  export * from '@udecode/plate-suggestion/react';
  export * from '@udecode/plate-tabbable/react';
  export * from '@udecode/plate-table/react';
  export * from '@udecode/plate-toggle/react';
  export * from '@udecode/plate-resizable';
  ```

  - Replace all `'@udecode/plate'` and `'@udecode/plate/react'` with `'@/plate'` in your codebase.

## 41.0.14

## 41.0.13

## 41.0.12

## 41.0.10

## 41.0.9

## 41.0.8

## 41.0.7

## 41.0.6

## 41.0.5

## 41.0.2

## 41.0.0

## 40.3.1

## 40.2.8

## 40.2.7

## 40.0.3

## 40.0.2

## 40.0.1

## 40.0.0

## 39.2.21

## 39.2.20

## 39.2.15

## 39.2.13

## 39.2.12

## 39.2.1

## 39.1.8

## 39.1.4

## 39.1.3

## 39.0.0

## 38.0.6

## 38.0.4

## 38.0.3

## 38.0.2

## 38.0.1

## 38.0.0

## 37.0.8

## 37.0.7

## 37.0.5

## 37.0.4

## 37.0.3

## 37.0.0

### Major Changes

- [#3420](https://github.com/udecode/plate/pull/3420) by [@zbeyens](https://github.com/zbeyens) –
  - Split build into `@udecode/plate-common` and `@udecode/plate-common/react`.
  - NEW `/react` exports `@udecode/react-hotkeys`

## 36.3.9

## 36.3.7

## 36.3.4

## 36.2.1

## 36.0.6

## 36.0.3

## 36.0.0

## 35.3.2

## 34.0.5

## 34.0.4

## 34.0.2

## 34.0.1

## 34.0.0

## 33.0.4

### Patch Changes

- [#3199](https://github.com/udecode/plate/pull/3199) by [@zbeyens](https://github.com/zbeyens) – Fix `PlateElementProps` type

## 33.0.3

## 33.0.0

## 32.0.1

## 32.0.0

## 31.3.2

## 31.0.0

## 30.4.5

## 30.1.2

## 30.0.0

## 29.1.0

## 29.0.1

## 29.0.0

### Minor Changes

- [#2829](https://github.com/udecode/plate/pull/2829) by [@zbeyens](https://github.com/zbeyens) –
  - re-export `@udecode/react-utils`

### Patch Changes

- [#2829](https://github.com/udecode/plate/pull/2829) by [@zbeyens](https://github.com/zbeyens) –
  - Fix import from RSC

## 28.0.0

## 27.0.3

## 27.0.0

## 25.0.1

## 25.0.0

## 24.5.2

## 24.4.0

### Minor Changes

- [#2675](https://github.com/udecode/plate/pull/2675) by [@zbeyens](https://github.com/zbeyens) – Support slate-react 0.99.0

## 24.3.6

## 24.3.5

## 24.3.2

## 24.3.1

## 24.3.0

## 24.2.0

## 24.0.2

## 24.0.1

## 24.0.0

## 23.7.4

## 23.7.0

## 23.6.0

## 23.3.1

## 23.3.0

## 22.0.2

## 22.0.1

## 22.0.0

## 21.5.0

## 21.4.2

## 21.4.1

## 21.3.2

## 21.3.0

## 21.1.5

## 21.0.0

## 20.7.2

## 20.7.0

## 20.4.0

## 20.3.2

## 20.0.0
