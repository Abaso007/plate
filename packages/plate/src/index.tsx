export * from '@platejs/core';

export * from '@platejs/utils';

export * from '@platejs/slate';

export * from '@udecode/utils';
