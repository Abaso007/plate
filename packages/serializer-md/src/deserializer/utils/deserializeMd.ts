import type { SlateEditor } from '@udecode/plate-common';

import markdown from 'remark-parse';
import unified from 'unified';

import {
  type RemarkPluginOptions,
  remarkPlugin,
} from '../../remark-slate/index';
import { DeserializeMdPlugin } from '../DeserializeMdPlugin';

/** Deserialize content from Markdown format to Slate format. `editor` needs */
export const deserializeMd = (editor: SlateEditor, data: string) => {
  const { elementRules, indentList, textRules } =
    editor.getOptions(DeserializeMdPlugin);

  const tree: any = unified()
    .use(markdown)
    .use(remarkPlugin, {
      editor,
      elementRules,
      indentList,
      textRules,
    } as unknown as RemarkPluginOptions)
    .processSync(data);

  return tree.result;
};
