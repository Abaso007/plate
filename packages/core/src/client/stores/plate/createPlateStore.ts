import React from 'react';

import type { Value } from '@udecode/slate';
import type { JotaiStore } from 'jotai-x';

import { atom, createStore } from 'jotai';

import type { PlateEditor, PlateStoreState } from '../../../shared/types';

import { createAtomStore } from '../../libs';
import { createPlateFallbackEditor } from '../../utils';
import {
  usePlateControllerEditorStore,
  usePlateControllerExists,
} from '../plate-controller';

/**
 * A unique id used as a provider scope. Use it if you have multiple `Plate` in
 * the same React tree.
 *
 * @default PLATE_SCOPE
 */
export type PlateId = string;

export const PLATE_SCOPE = 'plate';

export const GLOBAL_PLATE_SCOPE = Symbol('global-plate');

export const createPlateStore = <
  V extends Value = Value,
  E extends PlateEditor<V> = PlateEditor<V>,
>({
  decorate = null,
  editor = createPlateFallbackEditor<V, E>(),
  id,
  isMounted = false,
  onChange = null,
  onSelectionChange = null,
  onValueChange = null,
  primary = true,
  readOnly = null,
  renderElement = null,
  renderLeaf = null,
  value = null as any,
  versionDecorate = 1,
  versionEditor = 1,
  versionSelection = 1,
  ...state
}: Partial<PlateStoreState<V, E>> = {}) =>
  createAtomStore(
    {
      decorate,
      editor,
      isMounted,
      onChange,
      onSelectionChange,
      onValueChange,
      primary,
      readOnly,
      renderElement,
      renderLeaf,
      value,
      versionDecorate,
      versionEditor,
      versionSelection,
      ...state,
    } as PlateStoreState<V, E>,
    {
      extend: (atoms) => ({
        trackedEditor: atom((get) => ({
          editor: get(atoms.editor),
          version: get(atoms.versionEditor),
        })),
        trackedSelection: atom((get) => ({
          selection: get(atoms.editor).selection,
          version: get(atoms.versionSelection),
        })),
      }),
      name: 'plate',
    }
  );

export const {
  PlateProvider: PlateStoreProvider,
  plateStore,
  usePlateStore,
} = createPlateStore();

export interface UsePlateEditorStoreOptions {
  debugHookName?: string;
}

export const usePlateEditorStore = (
  id?: PlateId,
  { debugHookName = 'usePlateEditorStore' }: UsePlateEditorStoreOptions = {}
): JotaiStore => {
  // Try to fetch the store from a Plate provider
  const localStore = usePlateStore(id).store({ warnIfNoStore: false }) ?? null;

  /**
   * To preserve hook order, only use `localStore` if it was present on first
   * render. This lets us call `usePlateControllerEditorStore` conditionally.
   */
  const [localStoreExists] = React.useState(!!localStore);

  // If no store was found, try to fetch the store from a PlateController
  const store = localStoreExists
    ? localStore
    : // eslint-disable-next-line react-hooks/rules-of-hooks
      usePlateControllerEditorStore(id);

  /**
   * If we still have no store, there are two possibilities.
   *
   * Case 1: There is neither a Plate nor a PlateController above us in the
   * tree. In this case, throw an error, since calling the hook will never
   * work.
   *
   * Case 2: There is a PlateController, but it has no active editor. In this
   * case, return a fallback store until an editor becomes active.
   */
  const plateControllerExists = usePlateControllerExists();
  const fallbackStore = React.useMemo(() => createStore(), []);

  if (!store) {
    if (plateControllerExists) {
      return fallbackStore;
    }

    throw new Error(
      `${debugHookName} must be used inside a Plate or PlateController`
    );
  }

  return store;
};

export const usePlateSelectors = (
  id?: PlateId,
  options?: UsePlateEditorStoreOptions
) => {
  const store = usePlateEditorStore(id, {
    debugHookName: 'usePlateSelectors',
    ...options,
  });

  return usePlateStore({ store }).get;
};

export const usePlateActions = (
  id?: PlateId,
  options?: UsePlateEditorStoreOptions
) => {
  const store = usePlateEditorStore(id, {
    debugHookName: 'usePlateActions',
    ...options,
  });

  return usePlateStore({ store }).set;
};

export const usePlateStates = (
  id?: PlateId,
  options?: UsePlateEditorStoreOptions
) => {
  const store = usePlateEditorStore(id, {
    debugHookName: 'usePlateStates',
    ...options,
  });

  return usePlateStore({ store }).use;
};

/** Get the closest `Plate` id. */
export const usePlateId = (): PlateId =>
  usePlateSelectors(undefined, { debugHookName: 'usePlateId' }).editor().id;
