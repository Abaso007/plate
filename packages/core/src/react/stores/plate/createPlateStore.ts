import React from 'react';

import { atom } from 'jotai';

import type { PlateEditor } from '../../editor/PlateEditor';
import type { PlateChangeKey, PlateStoreState } from './PlateStore';

import { createAtomStore } from '../../libs';
import { createPlateFallbackEditor } from '../../utils';
import {
  usePlateControllerExists,
  usePlateControllerStore,
} from '../plate-controller';

export const PLATE_SCOPE = 'plate';

export const GLOBAL_PLATE_SCOPE = Symbol('global-plate');

export const createPlateStore = <E extends PlateEditor = PlateEditor>({
  id,
  containerRef = { current: null },
  decorate = null,
  editor,
  isMounted = false,
  primary = true,
  readOnly = null,
  renderElement = null,
  renderLeaf = null,
  scrollRef = { current: null },
  versionDecorate = 1,
  versionEditor = 1,
  versionSelection = 1,
  versionValue = 1,
  onChange = null,
  onSelectionChange = null,
  onValueChange = null,
  ...state
}: Partial<PlateStoreState<E>> = {}) =>
  createAtomStore(
    {
      containerRef,
      decorate,
      editor,
      isMounted,
      primary,
      readOnly,
      renderElement,
      renderLeaf,
      scrollRef,
      versionDecorate,
      versionEditor,
      versionSelection,
      versionValue,
      onChange,
      onSelectionChange,
      onValueChange,
      ...state,
    } as PlateStoreState<E>,
    {
      name: 'plate',
      extend: (atoms) => ({
        trackedEditor: atom((get) => ({
          editor: get(atoms.editor),
          version: get(atoms.versionEditor),
        })),
        trackedSelection: atom((get) => ({
          selection: get(atoms.editor).selection,
          version: get(atoms.versionSelection),
        })),
        trackedValue: atom((get) => ({
          value: get(atoms.editor).children,
          version: get(atoms.versionValue),
        })),
      }),
    }
  );

export const {
  PlateProvider: PlateStoreProvider,
  plateStore,
  usePlateStore: usePlateLocalStore,
} = createPlateStore();

export const usePlateStore = (id?: string) => {
  // Try to fetch the store from a Plate provider
  const localStore =
    usePlateLocalStore({ scope: id, warnIfNoStore: false }) ?? null;

  const [localStoreExists] = React.useState(!!localStore.store);

  // If no store was found, try to fetch the store from a PlateController
  const store = (
    localStoreExists
      ? localStore
      : // eslint-disable-next-line react-hooks/rules-of-hooks
        usePlateControllerStore(id)
  ) as typeof localStore;

  /**
   * If we still have no store, there are two possibilities.
   *
   * Case 1: There is neither a Plate nor a PlateController above us in the
   * tree. In this case, throw an error, since calling the hook will never
   * work.
   *
   * Case 2: There is a PlateController, but it has no active editor. In this
   * case, return a fallback store until an editor becomes active.
   */
  const plateControllerExists = usePlateControllerExists();

  const fallbackStore = React.useRef<typeof localStore>(undefined);

  if (!store) {
    if (plateControllerExists) {
      if (!fallbackStore.current) {
        fallbackStore.current =
          createPlateStore().usePlateStore() as unknown as typeof localStore;
      }
      return fallbackStore.current!;
    }

    throw new Error(
      `Plate hooks must be used inside a Plate or PlateController`
    );
  }
  return store;
};

// ─── Selectors ───────────────────────────────────────────────────────────────

/** Get the closest `Plate` id. */
export const useEditorId = (): string => usePlateStore().useEditorValue().id;

export const useEditorContainerRef = (id?: string) => {
  return usePlateStore(id).useContainerRefValue();
};

export const useEditorScrollRef = (id?: string) => {
  return usePlateStore(id).useScrollRefValue();
};

/** Returns the scrollRef if it exists, otherwise returns the containerRef. */
export const useScrollRef = (id?: string) => {
  const scrollRef = useEditorScrollRef(id);
  const containerRef = useEditorContainerRef(id);

  return scrollRef.current ? scrollRef : containerRef;
};

export const useEditorMounted = (id?: string): boolean => {
  return !!usePlateStore(id).useIsMountedValue();
};

/**
 * Whether the editor is read-only. You can also use `useReadOnly` from
 * `slate-react` in node components.
 */
export const useEditorReadOnly = (id?: string): boolean => {
  return !!usePlateStore(id).useReadOnlyValue();
};

/** Get editor ref which is never updated. */
export const useEditorRef = <E extends PlateEditor = PlateEditor>(
  id?: string
): E => {
  return (
    (usePlateStore(id).useEditorValue() as E) ?? createPlateFallbackEditor()
  );
};

/** Get the editor selection (deeply memoized). */
export const useEditorSelection = (id?: string) =>
  usePlateStore(id).useTrackedSelectionValue().selection;

/** Get editor state which is updated on editor change. */
export const useEditorState = <E extends PlateEditor = PlateEditor>(
  id?: string
): E => {
  return usePlateStore(id).useTrackedEditorValue().editor;
};

/** Version incremented on each editor change. */
export const useEditorVersion = (id?: string) => {
  return usePlateStore(id).useVersionValueValue();
};

/** Version incremented on selection change. */
export const useSelectionVersion = (id?: string) => {
  return usePlateStore(id).useVersionSelectionValue();
};

/** Get the editor value (deeply memoized). */
export const useEditorValue = (id?: string) =>
  usePlateStore(id).useTrackedValueValue().value;

/** Version incremented on value change. */
export const useValueVersion = (id?: string) => {
  return usePlateStore(id).useVersionValueValue();
};

// ─── Actions ─────────────────────────────────────────────────────────────────

export const useIncrementVersion = (key: PlateChangeKey, id?: string) => {
  const previousVersionRef = React.useRef(1);

  const store = usePlateStore(id);

  const setVersionDecorate = store.useSetVersionDecorate();
  const setVersionSelection = store.useSetVersionSelection();
  const setVersionValue = store.useSetVersionEditor();

  return React.useCallback(() => {
    const nextVersion = previousVersionRef.current + 1;

    switch (key) {
      case 'versionDecorate': {
        setVersionDecorate(nextVersion);

        break;
      }
      case 'versionEditor': {
        setVersionValue(nextVersion);

        break;
      }
      case 'versionSelection': {
        setVersionSelection(nextVersion);

        break;
      }
      // No default
    }

    previousVersionRef.current = nextVersion;
  }, [key, setVersionDecorate, setVersionSelection, setVersionValue]);
};

export const useRedecorate = (id?: string) => {
  const updateDecorate = useIncrementVersion('versionDecorate', id);

  return React.useCallback(() => {
    updateDecorate();
  }, [updateDecorate]);
};
