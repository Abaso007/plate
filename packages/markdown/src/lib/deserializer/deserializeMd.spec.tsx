/** @jsx jsxt */

import { createSlateEditor } from '@udecode/plate';
import { BaseHorizontalRulePlugin } from '@udecode/plate-horizontal-rule';
import { jsxt } from '@udecode/plate-test-utils';

import { deserializeMd } from './deserializeMd';

const editor = createSlateEditor();

jsxt;

describe('deserializeMd', () => {
  it('should deserialize paragraph', () => {
    const result = deserializeMd(editor, 'Hello, world!');

    const output = (
      <editor>
        <hp>Hello, world!</hp>
      </editor>
    ) as any;

    expect(result).toEqual(output.children);
  });

  it('should deserialize link', () => {
    const result = deserializeMd(
      editor,
      '[Hello, world!](https://example.com)'
    );

    const output = (
      <editor>
        <hp>
          <ha url="https://example.com">Hello, world!</ha>
        </hp>
      </editor>
    ) as any;

    expect(result).toEqual(output.children);
  });

  it('should deserialize blockquote', () => {
    const result = deserializeMd(editor, '> Hello, world!');

    const output = (
      <editor>
        <hblockquote>
          <hp>Hello, world!</hp>
        </hblockquote>
      </editor>
    ) as any;

    expect(result).toEqual(output.children);
  });

  it('should deserialize code blocks', () => {
    const input =
      '```\nCode block 1 line 1\nCode block 1 line 2\n```\n\n```\nCode block 2 line 1\n```';

    const output = (
      <fragment>
        <hcodeblock>
          <hcodeline>Code block 1 line 1</hcodeline>
          <hcodeline>Code block 1 line 2</hcodeline>
        </hcodeblock>
        <hcodeblock>
          <hcodeline>Code block 2 line 1</hcodeline>
        </hcodeblock>
      </fragment>
    );

    expect(deserializeMd(editor, input)).toEqual(output);
  });

  it('should deserialize headings', () => {
    const input = Array.from(
      { length: 6 },
      (_, i) => `${'#'.repeat(i + 1)} Heading ${i + 1}`
    ).join('\n\n');

    const output = (
      <fragment>
        <hh1>Heading 1</hh1>
        <hh2>Heading 2</hh2>
        <hh3>Heading 3</hh3>
        <hh4>Heading 4</hh4>
        <hh5>Heading 5</hh5>
        <hh6>Heading 6</hh6>
      </fragment>
    );

    expect(deserializeMd(editor, input)).toEqual(output);
  });

  it('should deserialize hr', () => {
    const input = 'Line 1\n\n---\n\nLine 2';

    const output = (
      <fragment>
        <hp>Line 1</hp>
        <element type={BaseHorizontalRulePlugin.key}>
          <htext />
        </element>
        <hp>Line 2</hp>
      </fragment>
    );

    expect(deserializeMd(editor, input)).toEqual(output);
  });

  it('should deserialize images', () => {
    const input = 'No ![inline](https://example.com/example.png) images';

    const output = (
      <fragment>
        <hp>No </hp>
        <himg
          caption={
            <fragment>
              <htext>inline</htext>
            </fragment>
          }
          url="https://example.com/example.png"
        >
          <htext />
        </himg>
        <hp> images</hp>
      </fragment>
    );

    expect(deserializeMd(editor, input)).toEqual(output);
  });
});

describe('deserializeMd equation', () => {
  it('should deserialize block equation', () => {
    const input = `$$
\\frac{-b \\pm \\sqrt{b^2 - 4ac}}{2a}
$$`;

    const output = (
      <fragment>
        <hequation texExpression="\frac{-b \pm \sqrt{b^2 - 4ac}}{2a}">
          <htext />
        </hequation>
      </fragment>
    );

    expect(deserializeMd(editor, input)).toEqual(output);
  });

  it('should deserialize inline equation', () => {
    const input = `$E=mc^2$ `;

    const output = (
      <fragment>
        <hp>
          <hinlineequation texExpression="E=mc^2">
            <htext />
          </hinlineequation>
        </hp>
      </fragment>
    );

    expect(deserializeMd(editor, input)).toEqual(output);
  });
});
