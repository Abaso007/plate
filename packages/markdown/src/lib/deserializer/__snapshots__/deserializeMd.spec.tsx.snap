// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixures should deserialize image within list 1`] = `
[
  {
    "caption": [
      {
        "text": "alt text",
      },
    ],
    "children": [
      {
        "text": "",
      },
    ],
    "indent": 1,
    "listStyleType": "disc",
    "type": "img",
    "url": "https://example.com/image.png",
  },
]
`;

exports[`fixures when deserializing a empty value 1`] = `[]`;
